tests/sources/helpers/rest_client/test_client.py

```diff
@@ -163,6 +225,67 @@ def test_api_key_auth_success(self, rest_client: RESTClient):
         assert response.status_code == 200
         assert response.json()["data"][0] == {"id": 0, "title": "Post 0"}
 
+    def test_oauth2_client_credentials_flow_auth_success(self, rest_client_oauth: RESTClient):
+        response = rest_client_oauth.get("/protected/posts/bearer-token")
+
+        assert response.status_code == 200
+        assert "test-token" in response.request.headers["Authorization"]
+
+        pages_iter = rest_client_oauth.paginate("/protected/posts/bearer-token")
+
+        assert_pagination(list(pages_iter))
+
+    def test_oauth2_client_credentials_flow_wrong_client_id(self, rest_client: RESTClient):
+        auth = OAuth2ClientCredentialsExample(
+            access_token_request_data={
+                "grant_type": "client_credentials",
+            },
+            client_id=cast(TSecretStrValue, "invalid-client-id"),
+            client_secret=cast(TSecretStrValue, "test-client-secret"),
+        )
+
+        with pytest.raises(HTTPError) as e:
+            rest_client.get(
+                "/protected/posts/bearer-token",
+                auth=auth,
+            )
+        assert e.type == HTTPError
+        assert e.match("401 Client Error")
+
+    def test_oauth2_client_credentials_flow_wrong_client_secret(self, rest_client: RESTClient):
+        auth = OAuth2ClientCredentialsExample(
+            access_token_request_data={
+                "grant_type": "client_credentials",
+            },
+            client_id=cast(TSecretStrValue, "test-client-id"),
+            client_secret=cast(TSecretStrValue, "invalid-client-secret"),
+        )
+
+        with pytest.raises(HTTPError) as e:
+            rest_client.get(
+                "/protected/posts/bearer-token",
+                auth=auth,
+            )
+        assert e.type == HTTPError
+        assert e.match("401 Client Error")
+
+    def test_oauth_token_expired_refresh(self, rest_client_immediate_oauth_expiry: RESTClient):
+        rest_client = rest_client_immediate_oauth_expiry
+        assert rest_client.auth.access_token is None
+        response = rest_client.get("/protected/posts/bearer-token")
+        assert response.status_code == 200
+        assert rest_client.auth.access_token is not None
+        expiry_0 = rest_client.auth.token_expiry
+        rest_client.auth.token_expiry = rest_client.auth.token_expiry.subtract(seconds=1)
```

Comment:
@burnash do you see a better way of testing a token refresh? An alternative might be to spy on the token refresh method and assert it got called. But maybe this integration test is more effective and it completes in milliseconds too.